gapic/src/main/com/google/gapid/views/DeviceDialog.java

```diff
@@ -0,0 +1,322 @@
+    synchronized (mutex) {
+      if (!dialogExists) {
+        dialog = new SelectReplayDeviceDialog(shell, models, widgets);
+        dialogExists = true;
+      }
+    }
+
+    if (dialog != null) {
+      // Skip dialog if there is only one compatible replay device,
+      // which is validated or validation-skipped.
+      boolean skipDialog = false;
+      Device.Instance device = null;
+      if (models.devices.getReplayDevices() != null
+          && models.devices.getReplayDevices().size() == 1) {
+        device = models.devices.getReplayDevices().get(0);
+        DeviceValidationResult result = models.devices.getValidationStatus(
+            new DeviceCaptureInfo(Paths.device(device.getID()), device, null, null));
+        skipDialog = result.passed || result.skipped;
+      }
+
+      if (skipDialog) {
+        models.devices.selectReplayDevice(device);
+      } else {
+        dialog.setBlockOnOpen(true);
+        dialog.open();
+      }
+    }
+
+    synchronized (mutex) {
+      if (dialog != null) {
+        dialogExists = false;
+      }
+    }
+  }
+
+  /**
+   * Dialog to select a replay device.
+   */
+  private static class SelectReplayDeviceDialog extends DialogBase {
+
+    private final Models models;
+    private final Widgets widgets;
+
+    private Label noCompatibleDeviceFound;
+    private ComboViewer deviceCombo;
+    private LoadingIndicator.Widget deviceLoader;
+    private LoadingIndicator.Widget validationStatusLoader;
+    private Link validationStatusText;
+    private boolean validationPassed;
+    private Button openTrace;
+
+    private final SingleInFlight rpcController = new SingleInFlight();
+
+    public SelectReplayDeviceDialog(Shell shell, Models models, Widgets widgets) {
+      super(shell, widgets.theme);
+      this.models = models;
+      this.widgets = widgets;
+      validationPassed = false;
+    }
+
+    @Override
+    public String getTitle() {
+      return Messages.SELECT_DEVICE_TITLE;
+    }
+
+    @Override
+    protected Control createDialogArea(Composite parent) {
+      Composite composite = (Composite) super.createDialogArea(parent);
+
+      if (!models.capture.isGraphics()) {
+        createLabel(composite,
+            "Error: no graphics capture loaded?"
+                + " This dialog should not be reachable without a graphics capture loaded.")
+                    .setForeground(theme.missingInput());
+        return composite;
+      }
+
+      // Recap capture info
+      createLabel(composite, "Capture name: " + models.capture.getName());
+      createLabel(composite,
+          "Capture device: " + Devices.getLabel(models.capture.getData().capture.getDevice()));
+
+      // Warning when no compatible device found
+      noCompatibleDeviceFound = createLabel(composite, Messages.SELECT_DEVICE_NO_COMPATIBLE_FOUND);
+      noCompatibleDeviceFound.setForeground(theme.missingInput());
+
+      // Mirror the device combo from TracerDialog
+      Group mainGroup =
```

Comment:
I don't think we need the group here.

Reply:
I'd like to keep it, in order to have the same look and feel between tracing and replay device selection.